src/lib/lib.ml

```diff
@@ -225,6 +225,60 @@ struct
     Buffer.contents buf
 end
 
+module Utf8 =
+struct
+  type t = int list
+  exception Utf8 = Wasm.Utf8.Utf8
+
+  let rec is_valid s = is_valid' (List.map Char.code (String.explode s))
+  and is_valid' = function
+    | [] -> true
+    | b1::bs when b1 < 0x80 ->
+      is_valid' bs
+    | b1::bs when b1 < 0xc2 -> false
+    | b1::b2::bs when b1 < 0xe0 ->
+      (b2 land 0xc0 = 0x80) && is_valid' bs
+    | b1::b2::b3::bs when b1 < 0xf0 ->
+      (b2 land 0xc0 = 0x80) && (b3 land 0xc0 = 0x80) && is_valid' bs
+    | b1::b2::b3::b4::bs when b1 < 0xf8 ->
+      (b2 land 0xc0 = 0x80) && (b3 land 0xc0 = 0x80) && (b4 land 0xc0 = 0x80) && is_valid' bs
+    | _ -> false
+  
+  let con b = if b land 0xc0 = 0x80 then b land 0x3f else raise Utf8
+  let code min n =
+    if n < min || (0xd800 <= n && n < 0xe000) || n >= 0x110000 then raise Utf8
+    else n
+
+  let rec decode s = decode' [] (List.map Char.code (String.explode s))
+  and decode' acc = function
+    | [] -> List.rev acc
+    | b1::bs when b1 < 0x80 ->
+      decode' (code 0x0 b1 :: acc) bs
+    | b1::bs when b1 < 0xc2 -> raise Utf8
```

Comment:
```suggestion
    | b1::bs when b1 < 0xc2 -> raise Utf8
```
Should c2 be c0? Compare with https://github.com/WebAssembly/spec/blob/036365a10e47ac1bec0151c148d0b334b3b2ef2b/interpreter/binary/utf8.ml#L35